ruby/lib/gilded_rose.rb

```diff
@@ -1,56 +1,72 @@
 class GildedRose
+  attr_reader :items
 
   def initialize(items)
     @items = items
   end
 
-  def update_quality()
-    @items.each do |item|
-      if item.name != "Aged Brie" and item.name != "Backstage passes to a TAFKAL80ETC concert"
-        if item.quality > 0
-          if item.name != "Sulfuras, Hand of Ragnaros"
-            item.quality = item.quality - 1
-          end
-        end
-      else
-        if item.quality < 50
-          item.quality = item.quality + 1
-          if item.name == "Backstage passes to a TAFKAL80ETC concert"
-            if item.sell_in < 11
-              if item.quality < 50
-                item.quality = item.quality + 1
-              end
-            end
-            if item.sell_in < 6
-              if item.quality < 50
-                item.quality = item.quality + 1
-              end
-            end
-          end
-        end
-      end
-      if item.name != "Sulfuras, Hand of Ragnaros"
-        item.sell_in = item.sell_in - 1
+  def update_quality
+    items.each do |item|
+      update_item(item)
+    end
+  end
+
+  def update_item(item)
+    if item.name == "Sulfuras, Hand of Ragnaros"
+    elsif item.name == "Aged Brie"
+      aged_brie(item)
+    elsif item.name == "Backstage passes to a TAFKAL80ETC concert"
+      backstage_passes(item)
+    else
+      normal_item(item)
+    end
+  end
+end
+
+private
+
+def aged_brie(item)
+  if item.quality < 50
+    item.quality = item.quality + 1
+  end
+  item.sell_in = item.sell_in - 1
+  if item.sell_in < 0
+    if item.quality < 50
+      item.quality = item.quality + 1
+    end
+  end 
+end
+
+def backstage_passes(item)
+  if item.quality < 50
+    item.quality = item.quality + 1
+    if item.sell_in < 11
+      if item.quality < 50
```

Comment:
Same comment here and throughout the file -- try to reduce the amount of nesting.  See if combining conditionals (ifs) works.  Do your tests still pass?

AND do some things really need to be nested?  Un-nest things that can be un-nested.  Consider which is easier to read to someone reviewing your code.

```ruby
if item.quality < 50
    item.quality = item.quality + 1
    if item.sell_in < 11 && item.quality < 50
        item.quality = item.quality + 1
    end
end
```
VS 

```ruby
if item.quality < 50 then item.quality = item.quality + 1 end
if item.sell_in < 11 && item.quality < 50 then item.quality = item.quality + 1 end
```

See https://stackoverflow.com/questions/3827874/one-line-if-statement-not-working

Here's another thought.  Once you do some refactoring based on the ideas above, look for repeated code throughout the methods and use the `extract method` refactoring pattern.  For example, this concept is repeated constantly:

```ruby
  if item.quality < 50
    item.quality = item.quality + 1
  end

```

Turn it into a method and have your other methods call it.

```
def checkForQualityIncrease(item)
  if item.quality < 50 then item.quality = item.quality + 1 end
end
```

Do this for similarly repeated concepts.  Extract them into their own method, and call this helper method where appropriate.

Reply:
created helper functions